{"pageProps":{"postData":{"contentHtml":"<p>Pythonで始める機械学習でお勉強中にFutureWarningに遭遇しました。GroupKFoldで起こったのですが、とりあえず今は使えるもののこの先使えなくなることが示されているので修正します。</p>\n<h2>FutureWarningを発生させてみる</h2>\n<p>少し前の書籍の型に合わせてコードを書いてみます。<del>FutureWarning</del>が発生するはずです。</p>\n<pre><code class=\"hljs language-python\"><span class=\"hljs-meta\">@python</span>\n\n<span class=\"hljs-keyword\">from</span> sklearn.model_selection <span class=\"hljs-keyword\">import</span> GroupKFold  <span class=\"hljs-comment\"># グループ交差検証</span>\n<span class=\"hljs-keyword\">from</span> sklearn.datasets <span class=\"hljs-keyword\">import</span> make_blobs  <span class=\"hljs-comment\"># 合成データ生成</span>\n\n<span class=\"hljs-comment\"># 合成データセットを生成</span>\nX, y = make_blobs(n_samples=<span class=\"hljs-number\">10</span>, random_state=<span class=\"hljs-number\">0</span>)\n<span class=\"hljs-comment\"># 最初の2サンプルが同じグループに、次の4つが同じグループに</span>\ngroups = [<span class=\"hljs-number\">0</span>, <span class=\"hljs-number\">0</span>, <span class=\"hljs-number\">1</span>, <span class=\"hljs-number\">1</span>, <span class=\"hljs-number\">1</span>, <span class=\"hljs-number\">1</span>, <span class=\"hljs-number\">3</span>, <span class=\"hljs-number\">3</span>, <span class=\"hljs-number\">4</span>, <span class=\"hljs-number\">4</span>]\nscores = cross_val_score(logreg, X, y, groups, cv=GroupKFold(n_splits=<span class=\"hljs-number\">3</span>))\n<span class=\"hljs-built_in\">print</span>(<span class=\"hljs-string\">\"Cross-validation scores:\\n{}\"</span>.<span class=\"hljs-built_in\">format</span>(scores))</code></pre>\n<p>おそらくこの型が少し前の技術書の仕様だと思います。<em>Pythonで始める機械学習</em>もこの型でした。</p>\n<p>結果をみてみます。</p>\n<pre><code class=\"hljs language-bash\">Cross-validation scores:\n[0.5 1.  1. ]\n/Users/user/opt/anaconda3/envs/ML/lib/python3.6/site-packages/sklearn/utils/validation.py:71: \nFutureWarning: Pass groups=[0, 0, 1, 1, 1, 1, 3, 3, 4, 4] as keyword args. \nFrom version 0.25 passing these as positional arguments will result <span class=\"hljs-keyword\">in</span> an error\nFutureWarning)</code></pre>\n<p>こんな感じのWarningが発生します。</p>\n<p>最後の方を読んでみると、<strong>バージョン0.25からはgroups配列は位置引数としてだとエラーになるよ</strong>みたいな感じのことを言っています。改善方法は至ってシンプルです。<br>\n次に示します。</p>\n<h2>改善策</h2>\n<p>改善策は至ってシンプルです。</p>\n<p>**groups=(グループ配列)**にすればいいのです。直したコードを示します。</p>\n<pre><code class=\"hljs language-python\"><span class=\"hljs-meta\">@python</span>\n\n<span class=\"hljs-keyword\">from</span> sklearn.model_selection <span class=\"hljs-keyword\">import</span> GroupKFold  <span class=\"hljs-comment\"># グループ交差検証</span>\n<span class=\"hljs-keyword\">from</span> sklearn.datasets <span class=\"hljs-keyword\">import</span> make_blobs  <span class=\"hljs-comment\"># 合成データ生成</span>\n\n<span class=\"hljs-comment\"># 合成データセットを生成</span>\nX, y = make_blobs(n_samples=<span class=\"hljs-number\">10</span>, random_state=<span class=\"hljs-number\">0</span>)\n<span class=\"hljs-comment\"># 最初の2サンプルが同じグループに、次の4つが同じグループに</span>\ngroups = [<span class=\"hljs-number\">0</span>, <span class=\"hljs-number\">0</span>, <span class=\"hljs-number\">1</span>, <span class=\"hljs-number\">1</span>, <span class=\"hljs-number\">1</span>, <span class=\"hljs-number\">1</span>, <span class=\"hljs-number\">3</span>, <span class=\"hljs-number\">3</span>, <span class=\"hljs-number\">4</span>, <span class=\"hljs-number\">4</span>]\nscores = cross_val_score(logreg, X, y, groups=groups, cv=GroupKFold(n_splits=<span class=\"hljs-number\">3</span>))\n<span class=\"hljs-built_in\">print</span>(<span class=\"hljs-string\">\"Cross-validation scores:\\n{}\"</span>.<span class=\"hljs-built_in\">format</span>(scores))</code></pre>\n<p>直したのはこの部分</p>\n<pre><code class=\"hljs language-python\">scores = cross_val_score(logreg, X, y, groups=groups, cv=GroupKFold(n_splits=<span class=\"hljs-number\">3</span>))</code></pre>\n<p>単に<em>groups=groups</em>にするだけです。これだけで直ります。</p>\n<p>やはりFutureWarningが多いですね〜この周りは。</p>\n","Title":"Scikit-learnのGroupKFoldでgroups配列でWarning発生!!","Date":"2020-08-18","Category":"Python","Tags":["Scikit-learn","Python"],"Authors":"ゆうぼう","Slug":"groupfold-groups","Thumbnail":"/images/thumbnails/network.jpg","Description":"Pythonで始める機械学習でお勉強中にFutureWarningに遭遇しました。GroupKFoldで起こったのですが、とりあえず今は使えるもののこの先使えなくなることが示されているので修正します。","Published":true},"categories":["Web","JavaScript","Competition","Cloud","Linux","Python","ML","Go","SQL"],"tags":["Apache","Appium","atmaCup","AWS","CentOS7","CentOS8","conda","CSS","ffmpeg","Flask","Go","HTML","JavaScript","JSON","Linux","Mac","make","map","MeCab","ML","MySQL","NLP","node.js","Pandas","Python","Scikit-learn","Selenium","subprocess","tensorflow","Tkinter","zsh","オブジェクト指向","デコレータ","データ分析","特殊メソッド"]},"__N_SSG":true}